### ISSUE TITLE
Support for LR-QAT

### ISSUE THREAD

**Issue by Juahyori** [pytorch/ao#2914]
Qualcomm research proposed a technique LR-QAT in their paper "Low-Rank Quantization-Aware Training for LLMs".

The core idea is that the low-rank weights are placed within the quantization grid of the model's weights using a custom downcasting operator.

The unique advantage of this is that it allows for a low rank adapter to control for the impact of quantization while still being absorbed into the main weights at inference, meaning that there's no inference overhead of the technique (and no lossy upcast to merge a LoRA adapter with, for example, NF4 weights in something like QLoRA).

Once a language model has been optimized under this framework, it's still suitable for further fine tuning, meaning that if one self distills a single target model using LR-QAT, it can be trained for a variety of downstream applications.

The memory use is quite favorable (relatively comparable to Q-LoRA), but has a variety of advantages to downstream inference usage.

Now, the good things out of the way, there's a few problems:

- The upcasting operator is a bit of development overhead. It requires a completely bespoke LoRA implementation that's not, to my eye, suitable for integration with existing tools.

- While a lot of logic is shared, I don't think the quantization grid logic will cleanly map into existing code.

- There's also some extra fixed point operators that are going to be a bit of a migraine to deal with.

- While memory-cheap, there is some computational overhead to the technique. I still think it's interesting, and has a lot of really favorable properties, but it's worth bearing in mind.

So, is there any possibility of or interest in adopting this technique within TorchAO? It's a fairly accessible recipe (particularly for end developers) and its inclusion could mean a fairly rich library of accessible model checkpoints up to about 24B parameters in size (as that's around the limit of what I think most developers will be able to optimize on GPUs at home), and my intuition is that even up to around 70B dense models should be accessible on fairly cheap GPU instances, as well.

So far as MoE, I think it'd take a lot of consideration because there's already a lot of ecosystem growing pains surrounding them (see: ongoing issues with expert dispatch in the Huggingface Transformers ecosystem which has been inherited by most finetuning frameworks with the notable exception of Torchtune), and many existing implementations have poor support / prospects for funky operators (particularly LoRA, etc).

**Comment by jerryzh168** [comment]
cc @andrewor14
